programs/mpl-staking/src/instructions/penalties/slash.rs

```diff
@@ -0,0 +1,196 @@
+    #[account(
+        mut,
+        seeds = [registrar.key().as_ref(), b"voter-weight-record".as_ref(), voter.load()?.voter_authority.key().as_ref()],
+        bump = voter.load()?.voter_weight_record_bump,
+        constraint = voter_weight_record.realm == registrar.load()?.realm,
+        constraint = voter_weight_record.governing_token_owner == voter.load()?.voter_authority,
+        constraint = voter_weight_record.governing_token_mint == registrar.load()?.realm_governing_token_mint,
+    )]
+    pub voter_weight_record: Account<'info, VoterWeightRecord>,
+
+    /// CHECK:
+    /// Ownership of the account will be checked in the rewards contract
+    /// It's the core account for the rewards contract, which will
+    /// keep track of all rewards and staking logic.
+    #[account(mut)]
+    pub reward_pool: UncheckedAccount<'info>,
+
+    /// CHECK: mining PDA will be checked in the rewards contract
+    /// PDA(["mining", mining owner <aka voter_authority in our case>, reward_pool],
+    /// reward_program)
+    #[account(mut)]
+    pub deposit_mining: UncheckedAccount<'info>,
+
+    /// CHECK: Rewards Program account
+    #[account(executable)]
+    pub rewards_program: UncheckedAccount<'info>,
+
+    pub token_program: Program<'info, Token>,
+}
+
+impl<'info> Slashing<'info> {
+    pub fn transfer_ctx(&self) -> CpiContext<'_, '_, '_, 'info, token::Transfer<'info>> {
+        let program = self.token_program.to_account_info();
+        let accounts = token::Transfer {
+            from: self.vault.to_account_info(),
+            to: self.realm_treasury.to_account_info(),
+            authority: self.voter.to_account_info(),
+        };
+        CpiContext::new(program, accounts)
+    }
+}
+
+/// Slashes the specified stake in transfers money to the treasury.
+///
+/// `deposit_entry_index`: The deposit entry to slash.
+/// `amount`: is in units of the native currency being slashed.
+/// `mining_owner`: The owner of the mining account.
+pub fn slash(
+    ctx: Context<Slashing>,
+    deposit_entry_index: u8,
+    amount: u64,
+    mining_owner: Pubkey,
+) -> Result<()> {
+    let registrar = ctx.accounts.registrar.load()?;
+
+    // this block is needed not to violate borrowing rules for the voter
+    {
+        require_keys_eq!(
+            registrar.realm_authority,
+            ctx.accounts.realm_authority.key(),
+            MplStakingError::InvalidRealmAuthority
+        );
+
+        let voter = &mut ctx.accounts.voter.load_mut()?;
+        // We should forbid slashes for votes that are in active votings state.
+        // Otherwise consequences are unpredictable.
+        // Not applicable for tokens that don't contribute to voting power.
+        let token_owner_record = load_token_owner_record(
+            &voter.voter_authority,
+            &ctx.accounts.token_owner_record.to_account_info(),
+            &registrar,
+        )?;
+        token_owner_record.assert_can_withdraw_governing_tokens()?;
```

Comment:
More as a mental exercise - wdyt of marking the slashed amount on the voter account and decreasing the rewards amounts and only transfer the slashed amount on withdrawal? This way we should keep the governance happy and be able to slash without any way to overcome it with voting.

Reply:
That's a good idea. 
However, we need to extend an account to keep track of how many tokens have been slashed.

Reply:
On the other hand, it means unless the user presses "withdraw", tokens will be on his stake forever. Probably, we don't want this. It's simpler to wait until the vote expires.